dev-tools/xtask/src/a4x2_package.rs

```diff
@@ -0,0 +1,382 @@
+    /// Bundle omicron live tests into the output tarball
+    #[clap(long)]
+    live_tests: bool,
+
+    /// Bundle omicron end-to-end-tests package into the output tarball
+    #[clap(long)]
+    end_to_end_tests: bool,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "kebab-case")]
+struct NextestConfig {
+    nextest_version: NextestConfigVersion,
+}
+
+#[derive(Deserialize, Debug)]
+struct NextestConfigVersion {
+    // required: String,
+    recommended: String,
+}
+
+struct Environment {
+    cargo: String,
+    home_dir: Utf8PathBuf,
+    work_dir: Utf8PathBuf,
+    src_dir: Utf8PathBuf,
+    out_dir: Utf8PathBuf,
+    omicron_dir: Utf8PathBuf,
+    in_ci: bool,
+}
+
+pub fn run_cmd(args: A4x2PackageArgs) -> Result<()> {
+    let sh = Shell::new()?;
+
+    let env = {
+        // Buildomat sets the CI environment variable. We'll work a bit
+        // differently in that case (no cleanup, use /work)
+        let in_ci = env::var("CI").is_ok();
+
+        let cargo =
+            std::env::var("CARGO").unwrap_or_else(|_| String::from("cargo"));
+
+        let home_dir = Utf8PathBuf::from(std::env::var("HOME")?);
+        let work_dir = if in_ci {
+            Utf8PathBuf::from("/work/a4x2-package")
+        } else {
+            home_dir.join(".cache/a4x2-package")
+        };
+
+        let src_dir = Utf8PathBuf::try_from(env::current_dir()?)?;
+
+        // Delete any results from previous runs. We don't mind if there's errors.
+        fs::remove_dir_all(&work_dir).ok();
+        // Create work dir
+        fs::create_dir_all(&work_dir)?;
+        sh.change_dir(&work_dir);
+
+        // Output. Maybe in CI we want this to be /out
+        let out_dir = work_dir.join("a4x2-package-out");
+        fs::create_dir_all(&out_dir)?;
+
+        // Clone of omicron that we can modify without causing trouble for anyone
+        let omicron_dir = work_dir.join("omicron");
+        fs::create_dir_all(&omicron_dir)?;
+
+        Environment {
+            cargo,
+            home_dir,
+            work_dir,
+            src_dir,
+            out_dir,
+            omicron_dir,
+            in_ci,
+        }
+    };
+
+    prepare_source(&sh, &env)?;
+
+    if args.end_to_end_tests {
+        build_end_to_end_tests(&sh, &env)?;
+    }
+
+    if args.live_tests {
+        build_live_tests(&sh, &env)?;
+    }
+
+    // This needs to happen last because it messes with the working tree in a
+    // way that end-to-end tests doesnt like when building
+    build_a4x2(&sh, &env)?;
+
+    create_output_artifact(&sh, &env)?;
+
+    Ok(())
+}
+
+/// Clone local working tree into work dir. This is nabbed from
+/// buildomat-at-home
+fn prepare_source(sh: &Shell, env: &Environment) -> Result<()> {
+    cmd!(sh, "banner 'prepare'").run()?;
+
+    // Get a ref we can checkout from the local working tree
+    let treeish = {
+        let _popdir = sh.push_dir(&env.src_dir);
+        let mut treeish = cmd!(sh, "git stash create").read()?;
```

Comment:
This is where I'd be inclined to:

- check if the working tree is dirty
- if so: bail unless some `--allow-dirty` is specified
- regardless, use `HEAD`

Reply:
Also:

* Use `$GIT` before `git`.
* Altering stash-related state is a visible operation and I don't think we should do it -- instead, there are ways to create a "background" commit from the current state.
* What happens if this is run in a non-git repo, e.g. a non-colocated Jujutsu repo?

Another option that has somewhat less source control interaction is:

* get the list of all tracked files in the repo ([here's instructions on how](https://github.com/camino-rs/paths0/wiki/Source-control-command-list:-git,-hg,-jj))
* create hardlinks for all of them in a mirrored structure. (Hardlinks are fine because basically all editors write files atomically, which breaks them.)

Reply:
> Altering stash-related state is a visible operation and I don't think we should do it

`git stash create` (not to be confused with `git stash push`) doesn't put the created commit object anywhere in the ref namespace, per the man page -- is it visible in some other way that creating a tree with `git update-index` etc isn't?